types/dispatcher.d.ts

```diff
@@ -232,6 +241,28 @@ declare namespace Dispatcher {
     onComplete?(trailers: string[] | null): void;
     /** Invoked when a body chunk is sent to the server. May be invoked multiple times for chunked requests */
     onBodySent?(chunkSize: number, totalBytesSent: number): void;
+
+    // New API
+
+    /** Invoked after request is starting to be processed */
+    onRequestStart?(/* controller: Controller */): void;
+    /** Invoked after headers data is sent */
+    // onRequestHeaders?(headers: Record<string, string>): void;
+    /** Invoked after payload data is sent. */
+    onRequestData?(chunk: Buffer | string): void;
+    /** Invoked after request has finished sending */
+    onRequestEnd?(): void;
+
+    /** Invoked after response is starting to be processed */
+    onResponseStart?(controller: Controller): void;
```

Comment:
I think you're not passing a controller here.
Also, in the `onRequestStart` above, you don't pass it so I would remove the comment at all.

Reply:
So we have 2 different controllers, one for the request and one for the response, they may run in parallel.